### PR TITLE
Abstract implementation of the User Profile Management Protocol

### PR DESCRIPTION
**Pull request by leonardr**
This branch implements and tests a simple implementation of https://github.com/NYPL-Simplified/Simplified/wiki/User-Profile-Management-Protocol that uses a swappable data store.

The data store used for testing keeps the read-only items in one dictionary and the writable items in another dictionary.

If there are any discrepancies between this and the spec, it's possible the spec needs to change, but more likely it's a bug in the implementation.

### CODE REVIEW

File: user_profile.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    MEDIA_TYPE = "vnd.librarysimplified/user-profile+json"
+    LINK_RELATION = "http://librarysimplified.org/terms/rel/user-profile"
+    SETTINGS_KEY = 'settings'
+    
+    def __init__(self, store):
+        """Constructor.
+
+        :param store: An instance of ProfileStore.
+        """
+        self.store = store
```
Comment (daryanypl) [inline user_profile.py]: is the store pointing to the profile server, or does it represent the library the user is signed up to?

Comment (leonardr) [inline user_profile.py]: It represents the user themselves. In my follow-up branch it's a wrapper around the `Patron` object.

File: user_profile.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        
+    def set(self, settable, full):
+        """(Try to) make the local store look like the provided Profile
+        document.
+
+        :param settable: The portion of the Profile document containing
+            settings that the client wants to change.
+
+        :param full: The full Profile document as provided by the client.
+            Should not be necessary but provided in case it's userful.
```
Comment (daryanypl) [inline user_profile.py]: useful

File: user_profile.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        """
+        raise NotImplementedError()
+    
+
+class DictionaryBasedProfileStore(object):
```
Comment (daryanypl) [inline user_profile.py]: inherit from ProfileStore

File: tests/test_user_profile.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        
+    def test_attempt_to_set_read_only_setting(self):
+        """You can't change the value of a setting that's not
+        writable.
+        """
+        headers = {"Content-Type" : ProfileController.MEDIA_TYPE}
+        body = json.dumps(dict(settings=dict(key="new value")))
+        problem = self.controller.put(headers, body)
+        eq_(400, problem.status_code)
+        eq_('"key" is not a writable setting.', problem.detail)
```
Comment (daryanypl) [inline tests/test_user_profile.py]: specific key?

Comment (leonardr) [inline tests/test_user_profile.py]: I'm not sure what this comment means. I'm testing that the name of the key ('key', defined on line 136) is used in the error message. If 'key' is a confusing name for a key I can change it.

File: user_profile.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+                if k not in writable:
+                    return INVALID_INPUT.detailed(
+                        _('"%s" is not a writable setting.' % k)
+                    )
+            try:
+                # Update the profile storage with the new settings.
+                self.storage.update(new_settings, profile_document)
+            except Exception, e:
+                # There was a problem updating the profile storage.
+                if hasattr(e, 'as_problem_detail_document'):
```
Comment (daryanypl) [inline user_profile.py]: I really like this as_problem_detail_document idea.

**Comment by daryanypl** [comment]
👍

**Pull request merged**